Add threading support for the `video` command

FreezyLemon opened the pull request:
The basic idea is that we hide the two decoders behind a `Mutex`, and let the worker threads decode two frames (one per decoder) to calculate the SSIMU2 score. It's written so that the Mutex lock is dropped as quickly as possible after decoding the frames, so another worker thread can start decoding ASAP. The resulting score value is sent via `result_tx` to `result_rx`, which is used in the main thread to accumulate the scores and handle the terminal output.

Some notes:
- I've tested a few approaches to multithreading, but it looks like there's always some memory bottleneck which is hit somewhere between 4 and 5 worker threads on my machine. I've "confirmed" it's a memory bottleneck by downclocking my RAM, which significantly reduced the performance ceiling.
- With every thread holding up to 2 frames at a time, it's very easy to hit OOM if you just enter a high `threads` value. I've thought about clamping the parameter to something like `(1, 20)` but then decided against it, as this is a user error.
- Minor point but I'm not sure if I should avoid using `drop` like that. Other languages with "automatic" memory management generally discourage manual deletion of variables. There's other ways to drop the value (move some of the code into a new scope, for example) but this seemed to be the easiest and most obvious way

shssoichiro commented on src/video.rs:
```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
+    }
+
+    // Needs to be dropped or the main thread never stops waiting for scores
+    drop(result_tx);
```
This is safe as far as I'm aware. It looks like it works because `drop` is just an empty function and the variable gets moved into it.

The other way to do this is to wrap your block of code where you want the variable to be alive into its own scope by using curly brackets. It's debatable which one is cleaner, I think the scoping technique is more common possibly.

FreezyLemon replied:
Mhm. I think doing it with a new scope is more common, too. But I like the explicit `drop` more in this case.

shssoichiro commented on src/main.rs:
```diff
@@ -40,6 +40,10 @@ enum Commands {
 
+        /// How many worker threads to use for decoding & calculating scores.
+        #[arg(long, short)]
+        threads: Option<usize>,
```
I want to add support for other methods of threading that don't add memory usage impact. So maybe we can call this parameter "frame_threads" or something to indicate that this is specifically the number of frames to run in parallel? And note in the help text that this will increase memory usage.

shssoichiro commented on src/video.rs:
```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
+                write!(w, "{:5.02} FPS", s.per_sec()).unwrap()
+            )
+            .with_key("pos", |s: &ProgressState, w: &mut dyn std::fmt::Write|
+                write!(w, "{:6}", s.pos()).unwrap()
```
I don't know that we gain anything from this (pos)

shssoichiro commented on src/video.rs:
```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
+            )
+            .unwrap()
+            .with_key("per_sec", |s: &ProgressState, w: &mut dyn std::fmt::Write|
+                write!(w, "{:5.02} FPS", s.per_sec()).unwrap()
```
Nit: I personally prefer lowercase "fps" but I don't really care that much in the end.

shssoichiro commented on src/video.rs:
```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
-        let result =
-            compute_frame_ssimulacra2(src_xyb, dst_xyb).expect("Failed to calculate ssimulacra2");
-        if verbose {
-            println!("Frame {frame}: {result:.8}");
```
Did this functionality disappear?

FreezyLemon replied:
Yes, I completely forgot about this. Added it back with the result ordering

```
...
[00:00:01] [12.21 fps] Frame 16
Frame 16: 93.59588478
[00:00:01] [15.97 fps] Frame 17
Frame 17: 93.46646854
[00:00:01] [16.67 fps] Frame 18
Frame 18: 93.50813436
[00:00:01] [13.54 fps] Frame 19
Frame 20: 93.67350370                  
...                                                         ...
```

shssoichiro commented on src/video.rs:
```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
+    };
+
+    for score in result_rx {
+        results.push(score);
```
Related to the verbose printing functionality, the channels will likely cause frame scores to be out of order. If we want to maintain frame order for both graphing and verbose printing to work correctly, we will probably have to associate each score with its frame number in a BTreeMap.

FreezyLemon commented:
Rebased & addressed the feedback. The Mutex now also guards a usize that is incremented every time both decoders successfully decode and return a frame to the score calculation.
